processor/batchprocessor/batch_processor.go

```diff
@@ -113,8 +113,16 @@ func (bp *batchProcessor) Shutdown(context.Context) error {
 
 func (bp *batchProcessor) startProcessingCycle() {
 	defer bp.goroutines.Done()
-	bp.timer = time.NewTimer(bp.timeout)
+	if bp.timeout != 0 && bp.sendBatchSize != 0 {
+		bp.timer = time.NewTimer(bp.timeout)
+	}
 	for {
+		// timerCh ensures we only block when there is a
+		// timer, since <- from a nil channel is blocking.
+		var timerCh <-chan time.Time
+		if bp.timer != nil {
+			timerCh = bp.timer.C
+		}
```

Comment:
Shouldn't this be done outside the for loop close where the `bp.timer` is initialized?

Reply:
Done.